server/hedley/modules/custom/hedley_admin/scripts/delete-duplicate-measurements.php

```diff
@@ -20,121 +22,78 @@
+  $deleted_for_encounter = 0;
+  $bundles = $fields["field_$encounter_type"]['bundles']['node'];
+  foreach ($bundles as $bundle) {
+    $query = db_select("field_data_field_$encounter_type", 'et');
+    $query->addField('et', "field_{$encounter_type}_target_id");
+    if ($encounter_type == 'session') {
+      $query->leftJoin('field_data_field_person', 'fp', 'fp.entity_id = et.entity_id');
+      $query->addField('fp', 'field_person_target_id');
     }
-
-    if (empty($summary[$person][$date][$type])) {
-      $summary[$person][$date][$type] = [];
+    $query->condition('et.bundle', $bundle);
+    $query->addExpression("COUNT(et.field_{$encounter_type}_target_id)", 'total');
+    $query->groupBy("et.field_{$encounter_type}_target_id");
+    if ($encounter_type == 'session') {
+      $query->groupBy("fp.field_person_target_id");
     }
-
-    $summary[$person][$date][$type][] = $node->nid;
-  }
-
-  $nid = end($ids);
-
-  if (round(memory_get_usage() / 1048576) >= $memory_limit) {
-    drush_print(dt('Stopped before out of memory. Start process from the node ID @nid', ['@nid' => $nid]));
-    return;
-  }
-}
-
-$count = 0;
-foreach (array_unique($people) as $person) {
-  foreach (array_unique($dates) as $date) {
-    foreach (array_keys($types) as $type) {
-      if (!empty($summary[$person][$date][$type])) {
-        $total = count($summary[$person][$date][$type]);
-        if ($total < 2) {
+    $query->havingCondition('total', 1, '>');
+    $query->range(0, 2000);
+    $duplicates_data = $query->execute()->fetchAll();
+
+    foreach ($duplicates_data as $data) {
+      $encounter = $data->{"field_{$encounter_type}_target_id"};
+      $query = db_select("field_data_field_$encounter_type", 'et');
+      $query->leftJoin('node', 'n', 'n.nid = et.entity_id');
+      $query->addField('et', 'entity_id');
+      if ($encounter_type == 'session') {
+        $person_id = $data->field_person_target_id;
+        if (!$person_id) {
+          // Can not resolve duplicates due to failure to retrieve person ID.
           continue;
         }
-
-        $last = array_pop($summary[$person][$date][$type]);
-        drush_print("Person $person, on $date, $type: Total - $total, last NID - $last");
-        $new_total = count($summary[$person][$date][$type]);
-        node_delete_multiple($summary[$person][$date][$type]);
-        drush_print("$new_total duplicates deleted");
-        $count += $new_total;
+        $query->leftJoin('field_data_field_person', 'fp', 'fp.entity_id = et.entity_id');
+        $query->addField('fp', 'field_person_target_id');
+        $query->condition('fp.field_person_target_id', $person_id);
       }
+      $query->condition('et.bundle', $bundle);
+      $query->condition("et.field_{$encounter_type}_target_id", $encounter);
+      $query->orderBy('n.vid', 'DESC');
+      $result = $query->execute()->fetchAllAssoc('entity_id');
+
+      $duplicates = array_keys($result);
+      $first = array_shift($duplicates);
```

Comment:
@anvmn I don't fully understand the logic behind this array_shift

Reply:
@mariano-dagostino 
The idea is that there are several nodes that are duplicates of each other. We want to delete all but one.
`$first = array_shift($duplicates);` pulls first node from that array and stores it, and `$duplicates` now got all remaining nodes, which are going to be deleted.

Reply:
@mariano-dagostino ^^